src/serializer/functions.js

```diff
@@ -388,14 +378,36 @@ export class Functions {
   }
 
   checkThatFunctionsAreIndependent(environmentRecordIdAfterGlobalCode: number) {
-    let additionalFunctions = this.__generateAdditionalFunctionsMap("__optimizedFunctions");
+    let additionalFunctionsToProcess = this.__generateInitialAdditionalFunctions("__optimizedFunctions");
+    let additionalFunctions = new Set(additionalFunctionsToProcess.map(entry => entry.value));
+    let optimizedFunctionsObject = this.moduleTracer.modules.logger.tryQuery(
+      () => Get(this.realm, this.realm.$GlobalObject, "__optimizedFunctions"),
+      this.realm.intrinsics.undefined
+    );
+    invariant(optimizedFunctionsObject instanceof ObjectValue);
 
-    for (let [funcValue] of additionalFunctions) {
+    while (additionalFunctionsToProcess.length > 0) {
+      let funcValue = additionalFunctionsToProcess.shift().value;
       invariant(funcValue instanceof ECMAScriptSourceFunctionValue);
       let call = this._callOfFunction(funcValue);
       let effects = this.realm.evaluatePure(() =>
         this.realm.evaluateForEffectsInGlobalEnv(call, undefined, "additional function")
       );
+      // look for newly registered optimized functions
+      let modifiedProperties = effects[3];
+      for (let [propertyBinding, newDescriptor] of modifiedProperties) {
+        if (propertyBinding.object === optimizedFunctionsObject) {
+          let newValue = newDescriptor && newDescriptor.value;
+          invariant(newValue);
+          let newEntry = this.__optimizedFunctionEntryOfValue(newValue);
```

Comment:
What about something like the following where the same function is requested to be optimized twice, in two slightly different settings? What do we want it to mean?

```js
(function() {
  var x = 23;
  function f() { return x; }
  function g() { x = 42; __optimize(f); return f; }
  function h() { __optimize(f); return f; }
  __optimize(g);
  __optimize(h);
  inspect = function() { return g() + "-" + h(); }
})();
```

Reply:
That's a good question, I think in general if we're doing nested additional functions, then we probably don't want to serialize multiple copies of the same additional function. 

We should also be able to deal with that case properly by figuring out we have to treat `x` as an abstract number instead of a concrete one, but this also becomes a question about what we want the behavior of additional functions capturing variables to be.

Reply:
How about for now if the same function is `__optimize`d twice, I just throw a `FatalError` saying it's not supported yet.

Reply:
If it were to be supported, how would it work? In general, how do we ensure that we optimize a function in the right setting?

Reply:
Well in general, we currently optimize everything in the global scope, so we don't do anything smart there. To support most nested optimized functions, we have to evaluate it in the context that it is defined or the context of one of the callsites. I think the easiest thing to do here would be to evaluate them with the effects of the first `__optimize` call we encounter.

Reply:
@cblappert: Failing is fine with me to start with.
@hermanventer: I don't like the idea of somewhat randomly picking one, this will lead to very confusing behavior.